src/Components/Components/src/RenderTree/EventHandlerDelegateWithClosuresComparer.cs

```diff
@@ -0,0 +1,130 @@
+            var newTargetType = right.Target!.GetType();
+
+            // if the types are not the same, or the methods are not the same, these possible closures are
+            // not the same anyway.
+            if (oldTargetType != newTargetType || left.Method != right.Method)
+            {
+                return false;
+            }
+
+            // for speed we cache the comparison functions per closure type.
+            var comparison = ClosureComparerCache.GetOrAdd(newTargetType, GetClosureComparerForType);
+            return comparison(left.Target, right.Target);
+        }
+
+        private static Func<object, object, bool> GetClosureComparerForType([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicFields)] Type type)
+        {
+            // if this type is not compiler generated, it is not a closure, and the previous tests where
+            // enough to conclude that these are not equal. We just cache a delegate to a function that returns false.
+            if (!type.CustomAttributes.Any(x => x.AttributeType == typeof(CompilerGeneratedAttribute)))
+            {
+                return (objA, objB) => false;
+            }
+
+            // build an array of lambdas that will get each public field value on the closure.
+            var fieldGetters = type.GetFields(
+                BindingFlags.Public |
+                BindingFlags.Instance |
+                BindingFlags.DeclaredOnly).Select(fieldInfo =>
+            {
+                var sourceParam = Expression.Parameter(typeof(object));
+                Expression returnExpression = Expression.Field(Expression.Convert(sourceParam, fieldInfo.DeclaringType!), fieldInfo);
+                if (fieldInfo.FieldType.IsValueType)
+                {
+                    // box if neccessary.
+                    returnExpression = Expression.Convert(returnExpression, typeof(object));
+                }
+                var lambda = Expression.Lambda(returnExpression, sourceParam);
+                return (Func<object?, object?>)lambda.Compile();
+            }).ToArray();
+
+            // the actual comparison function. It re-uses the array of public field 'getters' that we have built up
+            // before.
+            return (objA, objB) =>
+            {
+                // We have two instances of the same compiler generated class. Let's compare all public fields.
+                foreach (var fieldGetter in fieldGetters)
+                {
+                    var leftValue = fieldGetter(objA);
+                    var rightValue = fieldGetter(objB);
+
+                    // We might have recursive callbacks or delegates.
+                    // For example EventCallbackFactoryBinderExtensions.CreateBinderCore<T> wraps a delegate in a delegate
+                    // so this is not hypothetical.
+                    if (leftValue is EventCallback leftEventCallback && rightValue is EventCallback rightEventCallBack)
```

Comment:
Why do we do these type checks on every comparison, and not just once up front when generating the lambda? Does the capture object use `System.Object` as the type for its fields sometimes?

Reply:
Problem is that we don't know what the user will use as a type. Suppose the following situation:
```

@foreach (SomeClass item in items)
{
	object myDelegate = (SomeClass sc, MouseEventArgs args) => sc.DoSomething(args);
	<button @onclick="(args => ((Action<SomeClass, MouseEventArgs>)myDelegate(item, args))" />
}

```

Of course, this is insane and not code that anyone would write, but it does indicate that the declared type of a field on the closure does not necessarily have to match the runtime type. So `is` seems appropriate to me.